rust/src/dns/detect.rs

```diff
@@ -103,6 +121,363 @@ pub extern "C" fn SCDnsDetectRrtypeMatch(
+}
+
+unsafe extern "C" fn dns_query_get_data_wrapper(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int, local_id: u32,
+) -> *mut c_void {
+    return DetectHelperGetMultiData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        local_id,
+        dns_tx_get_query,
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn SCDetectDNSRegister() {
+    let kw = SCSigTableElmt {
+        name: b"dns.answer.name\0".as_ptr() as *const libc::c_char,
+        desc: b"DNS answer name sticky buffer\0".as_ptr() as *const libc::c_char,
+        url: b"/rules/dns-keywords.html#dns-answer-name\0".as_ptr() as *const libc::c_char,
+        Setup: dns_detect_answer_name_setup,
+        flags: SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER,
+        AppLayerTxMatch: None,
+        Free: None,
+    };
+    let _g_dns_answer_name_kw_id = DetectHelperKeywordRegister(&kw);
+    G_DNS_ANSWER_NAME_BUFFER_ID = DetectHelperMultiBufferMpmRegister(
+        b"dns.answer.name\0".as_ptr() as *const libc::c_char,
+        b"dns answer name\0".as_ptr() as *const libc::c_char,
+        ALPROTO_DNS,
+        true,
+        /* Register also in the TO_SERVER direction, even though this is not
+        normal, it could be provided as part of a request. */
+        true,
+        dns_answer_name_get_data_wrapper,
+    );
+    let kw = SCSigTableElmt {
+        name: b"dns.opcode\0".as_ptr() as *const libc::c_char,
+        desc: b"Match the DNS header opcode flag.\0".as_ptr() as *const libc::c_char,
+        url: b"rules/dns-keywords.html#dns-opcode\0".as_ptr() as *const libc::c_char,
+        AppLayerTxMatch: Some(dns_opcode_match),
+        Setup: dns_opcode_setup,
+        Free: Some(dns_opcode_free),
+        flags: 0,
+    };
+    G_DNS_OPCODE_KW_ID = DetectHelperKeywordRegister(&kw);
+    G_DNS_OPCODE_BUFFER_ID = DetectHelperBufferRegister(
+        b"dns.opcode\0".as_ptr() as *const libc::c_char,
+        ALPROTO_DNS,
+        true,
+        true,
+    );
+    let kw = SCSigTableElmt {
+        name: b"dns.query.name\0".as_ptr() as *const libc::c_char,
+        desc: b"DNS query name sticky buffer\0".as_ptr() as *const libc::c_char,
+        url: b"/rules/dns-keywords.html#dns-query-name\0".as_ptr() as *const libc::c_char,
+        Setup: dns_detect_query_name_setup,
+        flags: SIGMATCH_NOOPT | SIGMATCH_INFO_STICKY_BUFFER,
+        AppLayerTxMatch: None,
+        Free: None,
+    };
+    let _g_dns_query_name_kw_id = DetectHelperKeywordRegister(&kw);
+    G_DNS_QUERY_NAME_BUFFER_ID = DetectHelperMultiBufferMpmRegister(
+        b"dns.query.name\0".as_ptr() as *const libc::c_char,
+        b"dns query name\0".as_ptr() as *const libc::c_char,
+        ALPROTO_DNS,
+        true,
+        /* Register in both directions as the query is usually echoed back
+        in the response. */
+        true,
+        dns_query_name_get_data_wrapper,
+    );
```

Comment:
This results in:
```
        DetectAppLayerMultiRegister(name, alproto, SIG_FLAG_TOSERVER, 0, GetData, 2, 0);
```
However, on the C side:
```
    DetectAppLayerMultiRegister(keyword, ALPROTO_DNS, SIG_FLAG_TOSERVER, 0, GetBuffer, 2, 1);
```
I haven't looked into it, but is min-progress not needed? Is 1 equivalent to 0?  Is 2 equivalent to 1?

Its not a 1:1 conversion.

Reply:
> Is 1 equivalent to 0?

yes here

2 is some default priority...

Reply:
it's a priority for the buffer for mpm selection, essentially telling the engine it should prefer a buffer with a lower value over a higher value. It's not set very much I think. Would be good to turn it into a named enum in the whole API, so it's more clear what it is

Reply:
> Would be good to turn it into a named enum in the whole API, so it's more clear what it is

Should I create a ticket about it ?

Reply:
Good idea.

Reply:
https://redmine.openinfosecfoundation.org/issues/7563